proposals/0179-stricter-verification.md

```diff
@@ -0,0 +1,105 @@
+---
+simd: '0179'
+title: SBPF Stricter verification constraints
+authors:
+  - Alexander Meißner
+  - Lucas Steuernagel
+category: Standard
+type: Core
+status: Draft
+created: 2024-10-03
+---
+
+## Summary
+
+This SIMD proposes improvements in the verification rules of SBPF programs. 
+The introduction of the syscall instruction in SIMD-0178 allows for the 
+differentiation of internal and external calls, opening up for stricter 
+verification constraints and increased security of deployed programs.
+
+## Motivation
+
+The `call` instruction (opcode `0x85`) in SBF, especially when it requires a 
+relocation, is ambiguous. It does not carry information about whether it is 
+a system call or an internal call. This ambiguity prevents more rigid 
+verification constraints that would reject programs jumping to inconvenient 
+locations, like invalid destinations or the middle of an `lddw` instruction. 
+The new syscall instruction introduced in SIMD-0178 permits the 
+differentiation of internal and external calls, so we can introduce new 
+verification rules to prevent any unwarranted behavior.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+The following must go into effect if and only if a program indicates the SBF 
+version XX or higher in its ELF header e_flags field, according to the 
+specification of SIMD-0161.
+
+### Restrict functions’ last instruction
+
+Functions must only end with the `ja` (opcode `0x05`) or the exit (opcode 
+`0x9D` since SIMD-0178) instruction. Allowing calls to be the last instruction 
+of functions was inconvenient, because when the call returns, and there is no 
+other instruction to redirect the control flow, we will execute the very next 
+program counter, resulting in a fallthrough into another function’s code. 
+Offending this new validation condition must throw an 
+`VerifierError::InvalidFunction` error.
+
+### Restrict jump instruction destination
+
+All jump instructions, except for `call` (opcode `0x85`) and `callx` (opcode 
+`0x8D`), must now jump to a code location inside their own function. Jumping 
+to arbitrary locations hinders a precise program verification. 
+`VerifierError::JumpOutOfCode` must be thrown for offending this rule.
```

Comment:
How is this enforced? Is this enforced in the verifier or at runtime?

Reply:
This is a check at verification time (see error code) and is easy to enforce:
The verifier scans all instructions and keeps track of which function it is currently in, then checks jump targets against that range. The current function tracking can be done in constant time during the scan by checking the next symbol in the dynamic synbol table and then advancing that pointer once a function boundary is crossed.

Reply:
Could we add this description^ to the SIMD? 🙏

Reply:
Added.